File: listen-together-server/src/protocol.ts
/**
 * Wire protocol for Convx Listen Together.
 *
 * Mirrors app/src/main/kotlin/com/convx/music/listentogether/Protocol.kt.
 * Field names are snake_case because the Kotlin side declares @SerialName on
 * every multi-word field — changing a name here silently breaks decoding on
 * the client, which fails as a missing field rather than an error.
 *
 * The client starts in JSON, uncompressed, and only flips to protobuf+gzip if
 * the SERVER speaks protobuf first (ListenTogetherClient.kt:783-785). This
 * server is JSON-only and must stay that way, or every client permanently
 * switches format mid-session.
 */

export const C2S = {
  CREATE_ROOM: 'create_room',
  JOIN_ROOM: 'join_room',
  LEAVE_ROOM: 'leave_room',
  APPROVE_JOIN: 'approve_join',
  REJECT_JOIN: 'reject_join',
  PLAYBACK_ACTION: 'playback_action',
  BUFFER_READY: 'buffer_ready',
  KICK_USER: 'kick_user',
  TRANSFER_HOST: 'transfer_host',
  PING: 'ping',
  CHAT: 'chat',
  REQUEST_SYNC: 'request_sync',
  RECONNECT: 'reconnect',
  SUGGEST_TRACK: 'suggest_track',
  APPROVE_SUGGESTION: 'approve_suggestion',
  REJECT_SUGGESTION: 'reject_suggestion',
  // v2
  SET_CONTROL_MODE: 'set_control_mode',
  EXTEND_SESSION: 'extend_session',
} as const;

export const S2C = {
  ROOM_CREATED: 'room_created',
  JOIN_REQUEST: 'join_request',
  JOIN_APPROVED: 'join_approved',
  JOIN_REJECTED: 'join_rejected',
  USER_JOINED: 'user_joined',
  USER_LEFT: 'user_left',
  SYNC_PLAYBACK: 'sync_playback',
  BUFFER_WAIT: 'buffer_wait',
  BUFFER_COMPLETE: 'buffer_complete',
  ERROR: 'error',
  PONG: 'pong',
  HOST_CHANGED: 'host_changed',
  KICKED: 'kicked',
  SYNC_STATE: 'sync_state',
  RECONNECTED: 'reconnected',
  USER_RECONNECTED: 'user_reconnected',
  USER_DISCONNECTED: 'user_disconnected',
  SUGGESTION_RECEIVED: 'suggestion_received',
  SUGGESTION_APPROVED: 'suggestion_approved',
  SUGGESTION_REJECTED: 'suggestion_rejected',
  // v2
  CONTROL_MODE_CHANGED: 'control_mode_changed',
  ROOM_EXPIRING: 'room_expiring',
  ROOM_CLOSED: 'room_closed',
} as const;

export const PlaybackActions = {
  PLAY: 'play',
  PAUSE: 'pause',
  SEEK: 'seek',
  SKIP_NEXT: 'skip_next',
  SKIP_PREV: 'skip_prev',
  CHANGE_TRACK: 'change_track',
  QUEUE_ADD: 'queue_add',
  QUEUE_REMOVE: 'queue_remove',
  QUEUE_CLEAR: 'queue_clear',
  SYNC_QUEUE: 'sync_queue',
  SET_VOLUME: 'set_volume',
} as const;

/** Who may send playback actions. Owner-settable, enforced server-side. */
export type ControlMode = 'owner' | 'everyone';

export interface TrackInfo {
  id: string;
  title: string;
  artist: string;
  album?: string | null;
  duration: number;
  thumbnail?: string | null;
  suggested_by?: string | null;
}

export interface UserInfo {
  user_id: string;
  username: string;
  is_host: boolean;
  is_connected: boolean;
}

export interface RoomState {
  room_code: string;
  host_id: string;
  users: UserInfo[];
  current_track: TrackInfo | null;
  is_playing: boolean;
  position: number;
  last_update: number;
  volume: number;
  queue: TrackInfo[];
  // v2 additions. Older clients ignore unknown fields (the Kotlin Json is
  // configured with ignoreUnknownKeys), so adding these is backward safe.
  control_mode: ControlMode;
  expires_at: number;
  extensions_used: number;
}

export interface Message {
  type: string;
  payload?: unknown;
}

export function encode(type: string, payload?: unknown): string {
  return JSON.stringify(payload === undefined ? { type } : { type, payload });
}

File: app/src/main/kotlin/com/convx/music/listentogether/Protocol.kt
/**
 * Convx Project (C) 2026
 * Licensed under GPL-3.0 | See git history for contributors
 */

package com.convx.music.listentogether

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

/**
 * Message types for Listen Together protocol
 */
object MessageTypes {
    // Client -> Server
    const val CREATE_ROOM = "create_room"
    const val JOIN_ROOM = "join_room"
    const val LEAVE_ROOM = "leave_room"
    const val APPROVE_JOIN = "approve_join"
    const val REJECT_JOIN = "reject_join"
    const val PLAYBACK_ACTION = "playback_action"
    const val BUFFER_READY = "buffer_ready"
    const val KICK_USER = "kick_user"
    const val TRANSFER_HOST = "transfer_host"
    const val PING = "ping"
    const val CHAT = "chat"
    const val REQUEST_SYNC = "request_sync"
    const val RECONNECT = "reconnect"
    const val SUGGEST_TRACK = "suggest_track"
    const val APPROVE_SUGGESTION = "approve_suggestion"
    const val REJECT_SUGGESTION = "reject_suggestion"
    const val SET_CONTROL_MODE = "set_control_mode"
    const val EXTEND_SESSION = "extend_session"

    // Server -> Client
    const val ROOM_CREATED = "room_created"
    const val JOIN_REQUEST = "join_request"
    const val JOIN_APPROVED = "join_approved"
    const val JOIN_REJECTED = "join_rejected"
    const val USER_JOINED = "user_joined"
    const val USER_LEFT = "user_left"
    const val SYNC_PLAYBACK = "sync_playback"
    const val BUFFER_WAIT = "buffer_wait"
    const val BUFFER_COMPLETE = "buffer_complete"
    const val ERROR = "error"
    const val PONG = "pong"
    const val HOST_CHANGED = "host_changed"
    const val KICKED = "kicked"
    const val SYNC_STATE = "sync_state"
    const val RECONNECTED = "reconnected"
    const val USER_RECONNECTED = "user_reconnected"
    const val USER_DISCONNECTED = "user_disconnected"
    const val SUGGESTION_RECEIVED = "suggestion_received"
    const val SUGGESTION_APPROVED = "suggestion_approved"
    const val SUGGESTION_REJECTED = "suggestion_rejected"
    const val CONTROL_MODE_CHANGED = "control_mode_changed"
    const val ROOM_EXPIRING = "room_expiring"
    const val ROOM_CLOSED = "room_closed"
}

/** Room codes are not a fixed width across servers — the Cloudflare server
 *  mints six characters, older ones used eight — so the UI validates a range
 *  rather than an exact length. Matches the server's own `[A-Z0-9]{4,12}`. */
const val MIN_ROOM_CODE_LENGTH = 4
const val MAX_ROOM_CODE_LENGTH = 12

/** Who may send playback actions. Enforced by the server; the client only
 *  mirrors it so the transport controls can be disabled honestly. */
object ControlModes {
    const val OWNER = "owner"
    const val EVERYONE = "everyone"
}

/**
 * Playback action types
 */
object PlaybackActions {
    const val PLAY = "play"
    const val PAUSE = "pause"
    const val SEEK = "seek"
    const val SKIP_NEXT = "skip_next"
    const val SKIP_PREV = "skip_prev"
    const val CHANGE_TRACK = "change_track"
    const val QUEUE_ADD = "queue_add"
    const val QUEUE_REMOVE = "queue_remove"
    const val QUEUE_CLEAR = "queue_clear"
    const val SYNC_QUEUE = "sync_queue"
    const val SET_VOLUME = "set_volume"
}

/**
 * Base message structure
 */
@Serializable
data class Message(
    val type: String,
    val payload: JsonElement? = null
)

/**
 * Track information
 */
@Serializable
data class TrackInfo(
    val id: String,
    val title: String,
    val artist: String,
    val album: String? = null,
    val duration: Long, // milliseconds
    val thumbnail: String? = null,
    @SerialName("suggested_by") val suggestedBy: String? = null
)

/**
 * User information
 */
@Serializable
data class UserInfo(
    @SerialName("user_id") val userId: String,
    val username: String,
    @SerialName("is_host") val isHost: Boolean,
    @SerialName("is_connected") val isConnected: Boolean = true
)

/**
 * Room state
 */
@Serializable
data class RoomState(
    @SerialName("room_code") val roomCode: String,
    @SerialName("host_id") val hostId: String,
    val users: List<UserInfo>,
    @SerialName("current_track") val currentTrack: TrackInfo? = null,
    @SerialName("is_playing") val isPlaying: Boolean,
    val position: Long, // milliseconds
    @SerialName("last_update") val lastUpdate: Long, // unix timestamp ms
    val volume: Float = 1f,
    val queue: List<TrackInfo> = emptyList(),
    // v2. Defaulted so a room from an older server still decodes.
    @SerialName("control_mode") val controlMode: String = ControlModes.OWNER,
    /** Unix ms when the room closes; 0 when the server does not expire rooms. */
    @SerialName("expires_at") val expiresAt: Long = 0,
    @SerialName("extensions_used") val extensionsUsed: Int = 0
)

// Request payloads

@Serializable
data class CreateRoomPayload(
    val username: String
)

@Serializable
data class JoinRoomPayload(
    @SerialName("room_code") val roomCode: String,
    val username: String
)

@Serializable
data class ApproveJoinPayload(
    @SerialName("user_id") val userId: String
)

@Serializable
data class RejectJoinPayload(
    @SerialName("user_id") val userId: String,
    val reason: String? = null
)

@Serializable
data class PlaybackActionPayload(
    val action: String,
    @SerialName("track_id") val trackId: String? = null,
    val position: Long? = null, // milliseconds
    @SerialName("track_info") val trackInfo: TrackInfo? = null,
    @SerialName("insert_next") val insertNext: Boolean? = null,
    val queue: List<TrackInfo>? = null,
    @SerialName("queue_title") val queueTitle: String? = null,
    val volume: Float? = null,
    @SerialName("server_time") val serverTime: Long? = null,
    /** Stamped by the server on every broadcast. Lets a receiver tell its own
     *  action echoing back from someone else's, which is the difference between
     *  ignoring a loop and ignoring a real command. */
    @SerialName("from_user_id") val fromUserId: String? = null
)

@Serializable
data class BufferReadyPayload(
    @SerialName("track_id") val trackId: String
)

@Serializable
data class KickUserPayload(
    @SerialName("user_id") val userId: String,
    val reason: String? = null
)

@Serializable
data class TransferHostPayload(
    @SerialName("new_host_id") val newHostId: String
)

@Serializable
data class ChatPayload(
    val message: String,
    @SerialName("reply_to") val replyTo: RepliedMessage? = null
)

@Serializable
data class RepliedMessage(
    val username: String,
    val message: String
)

// Suggestions payloads

@Serializable
data class SuggestTrackPayload(
    @SerialName("track_info") val trackInfo: TrackInfo
)

@Serializable
data class SuggestionReceivedPayload(
    @SerialName("suggestion_id") val suggestionId: String,
    @SerialName("from_user_id") val fromUserId: String,
    @SerialName("from_username") val fromUsername: String,
    @SerialName("track_info") val trackInfo: TrackInfo
)

@Serializable
data class ApproveSuggestionPayload(
    @SerialName("suggestion_id") val suggestionId: String
)

@Serializable
data class RejectSuggestionPayload(
    @SerialName("suggestion_id") val suggestionId: String,
    val reason: String? = null
)

@Serializable
data class SuggestionApprovedPayload(
    @SerialName("suggestion_id") val suggestionId: String,
    @SerialName("track_info") val trackInfo: TrackInfo
)

@Serializable
data class SuggestionRejectedPayload(
    @SerialName("suggestion_id") val suggestionId: String,
    val reason: String? = null
)

// Response payloads

@Serializable
data class RoomCreatedPayload(
    @SerialName("room_code") val roomCode: String,
    @SerialName("user_id") val userId: String,
    @SerialName("session_token") val sessionToken: String,
    /** Null on pre-v2 servers, which sent no state to the creator. */
    val state: RoomState? = null
)

@Serializable
data class JoinRequestPayload(
    @SerialName("user_id") val userId: String,
    val username: String
)

@Serializable
data class JoinApprovedPayload(
    @SerialName("room_code") val roomCode: String,
    @SerialName("user_id") val userId: String,
    @SerialName("session_token") val sessionToken: String,
    val state: RoomState
)

@Serializable
data class JoinRejectedPayload(
    val reason: String
)

@Serializable
data class UserJoinedPayload(
    @SerialName("user_id") val userId: String,
    val username: String
)

@Serializable
data class UserLeftPayload(
    @SerialName("user_id") val userId: String,
    val username: String
)

@Serializable
data class BufferWaitPayload(
    @SerialName("track_id") val trackId: String,
    @SerialName("waiting_for") val waitingFor: List<String>
)

@Serializable
data class BufferCompletePayload(
    @SerialName("track_id") val trackId: String
)

@Serializable
data class ErrorPayload(
    val code: String,
    val message: String
)

@Serializable
data class ChatMessagePayload(
    @SerialName("user_id") val userId: String,
    val username: String,
    val message: String,
    val timestamp: Long,
    @SerialName("reply_to") val replyTo: RepliedMessage? = null
)

@Serializable
data class HostChangedPayload(
    @SerialName("new_host_id") val newHostId: String,
    @SerialName("new_host_name") val newHostName: String
)

@Serializable
data class KickedPayload(
    val reason: String
)

// v2 payloads

@Serializable
data class SetControlModePayload(
    @SerialName("control_mode") val controlMode: String
)

@Serializable
data class ControlModeChangedPayload(
    @SerialName("control_mode") val controlMode: String
)

@Serializable
data class RoomExpiringPayload(
    @SerialName("expires_at") val expiresAt: Long,
    @SerialName("extensions_left") val extensionsLeft: Int = 0
)

@Serializable
data class RoomClosedPayload(
    val reason: String? = null
)

/** Body of `POST /api/rooms`. The room code has to exist before the socket
 *  opens, because the server routes the connection by room. */
@Serializable
data class RoomAllocationResponse(
    @SerialName("room_code") val roomCode: String
)

/**
 * Sync state payload - sent to guest when they request current state
 */
@Serializable
data class SyncStatePayload(
    @SerialName("current_track") val currentTrack: TrackInfo?,
    @SerialName("is_playing") val isPlaying: Boolean,
    val position: Long,
    @SerialName("last_update") val lastUpdate: Long,
    val queue: List<TrackInfo>? = null,
    val volume: Float? = null,
    @SerialName("control_mode") val controlMode: String? = null,
    @SerialName("expires_at") val expiresAt: Long? = null
)

// Reconnection payloads

@Serializable
data class ReconnectPayload(
    @SerialName("session_token") val sessionToken: String
)

@Serializable
data class ReconnectedPayload(
    @SerialName("room_code") val roomCode: String,
    @SerialName("user_id") val userId: String,
    val state: RoomState,
    @SerialName("is_host") val isHost: Boolean
)

@Serializable
data class UserReconnectedPayload(
    @SerialName("user_id") val userId: String,
    val username: String
)

@Serializable
data class UserDisconnectedPayload(
    @SerialName("user_id") val userId: String,
    val username: String
)
